### PR TITLE
Joke refactor

### PR DESCRIPTION
**Pull request by prakritip0**
# Summary of change

* Refactored the code and functionality developed on joke branch
* Added a filter functionality to filter jokes by category within the json files

## Type of change

- [x] Bug fix
- [x] New feature

# How Has This Been Tested?

- [x] Manual testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/category.js [hunk 1]
```diff
@@ -2,29 +2,36 @@ import axios from "axios";
...
 
 const prompt = PromptSync();
 
 export const listCategories = async () => {
-    let categoryData = {};
+
+
     try {
-        categoryData = await axios.get("https://api.chucknorris.io/jokes/categories");
+        const categoryData = await axios.get("https://api.chucknorris.io/jokes/categories", { httpsAgent: new https.Agent({ keepAlive: true }) });
```
Comment (sarad-singh) [inline api/category.js]: Keep API links in constant as suggested before

File: api/category.js [hunk 2]
```diff
@@ -2,29 +2,36 @@ import axios from "axios";
...
-            finalData.category = categoryData.data[chosenCategory - 1];
-            return categoryData.data[chosenCategory - 1];
+            } else {
+                finalData.category = categoryData.data[chosenCategory - 1];
+                return categoryData.data[chosenCategory - 1];
 
+            }
+        } else {
+            console.log(chalk.red("invalid category!!!"))
+            return null
```
Comment (bisheshbh) [inline api/category.js]: We can use early return here

File: commands/commands.js [hunk 3]
```diff
@@ -35,37 +32,62 @@ export const askName = () => {
     return name;
 }
 
+const userDetailsArray = [];
 const storeData = () => {
-    const userDetails = JSON.stringify(finalData);
-    const readerFile = fs.existsSync(`${__dirname}/../files/${finalData.name}.json`, "utf-8");
+    userDetailsArray.length = 0
```
Comment (sarad-singh) [inline commands/commands.js:37]: Why is this line required?? as empty array's length is always 0

File: commands/commands.js [hunk 4]
```diff
@@ -35,37 +32,62 @@ export const askName = () => {
...
+    const checker = fs.existsSync(`${__dirname}/../files/${finalData.name}.json`)
+    if (checker) {
+        const content = fs.readFileSync(`${__dirname}/../files/${finalData.name}.json`, "utf-8")
+        const parsedContent = JSON.parse(content);
+        parsedContent.forEach(element => {
```
Comment (sarad-singh) [inline commands/commands.js:42]: See array destructuring, https://www.freecodecamp.org/news/array-destructuring-in-es6-30e398f21d10/
you could have used:
userDetailsArray = [...parsedContent]

File: commands/commands.js [hunk 5]
```diff
@@ -35,37 +32,62 @@ export const askName = () => {
...
+        });
 
-    if (!readerFile) {
-        fs.writeFileSync(`${__dirname}/../files/${finalData.name}.json`, userDetails, (err) => {
-            if (err) {
-                console.log("error")
-            }
-            console.log("\n" + "Your jokes are saved on " + chalk.blue(finalData.name) + chalk.blue(".json") + ". Please enter " + chalk.blue("view") + " to view the file.");
+        // userDetailsArray.push(content);
+        userDetailsArray.push(finalData);
```
Comment (sarad-singh) [inline commands/commands.js:47]: Why is this required???

File: commands/commands.js (hunk 3 above)
Comment (prakritip0) [reply commands/commands.js:37]: The array length is set to zero because everytime a user opts to hear more than 2 jokes, the jokes are doubled in the json file. It was done to erase the existing files in the array after each entry and push the existing file through content and new data through finalData

File: commands/commands.js (hunk 5 above)
Comment (prakritip0) [reply commands/commands.js:47]: Answered in one of your previous comments

File: api/category.js [hunk 6]
```diff
@@ -2,29 +2,36 @@ import axios from "axios";
...
+export const apiBaseURL = "https://api.chucknorris.io/jokes/";
 
 export const listCategories = async () => {
-    let categoryData = {};
+
     try {
-        categoryData = await axios.get("https://api.chucknorris.io/jokes/categories");
+        const categoryData = await axios.get(`${apiBaseURL}categories`, { httpsAgent: new https.Agent({ keepAlive: true }) });
         categoryData.data.forEach((category, i) => {
             console.log(chalk.green("[" + (i + 1) + "]") + " " + chalk.green(category));
```
Comment (saras-sth) [inline api/category.js]: suggestion: try to use template literals for situations like these:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: api/category.js [hunk 7]
```diff
@@ -2,29 +2,50 @@ import axios from "axios";
...
+            console.log(chalk.red("Invalid Category!!!"))
+            return
+        }
         if (isNaN(Number(chosenCategory))) {
             finalData.category = chosenCategory;
             return chosenCategory;
 
         } else {
             finalData.category = categoryData.data[chosenCategory - 1];
             return categoryData.data[chosenCategory - 1];
```
Comment (bisheshbh) [inline api/category.js]: We might not need else here

File: api/category.js [hunk 8]
```diff
@@ -2,29 +2,50 @@ import axios from "axios";
...
+            // console.log(chalk.green("[" + (i + 1) + "]") + " " + chalk.green(category));
+            console.log(chalk.green(`[${(i + 1)}] ${category}`));
         });
         console.log("\n" + chalk.bgGreen("Please enter one of the above categories to hear a joke!"));
 
         const chosenCategory = prompt(">");
+        if (chosenCategory == "") {
+            console.log(chalk.red("Invalid Category!!!"))
+            return
+        }
```
Comment (bisheshbh) [inline api/category.js:24]:
```suggestion
        if (!chosenCategory) {
            console.log(chalk.red("Invalid Category!!!"))
            return
        }
```
